persistence/docs/SAVEPOINTS_ROLLBACKS.md

```diff
@@ -0,0 +1,146 @@
+
+Even a single node is a small distributed system because it has multiple separate components that are communicating with each other (on the same machine or via unreliable networks).
+
+Since a node is part of a distributed network of nodes that are all trying to reach consensus on the "state of the world", we have to make sure that the data that we are storing is consistent internally in the first place.
+
+In the event of a failure at any level during this process, we cannot commit state non-atomically. That would make the node inconsistent and put it in a non-deterministic state that is not recoverable unless we have a way to rollback to a previous, clean, state with all the implications that this would have on the network. (social coordination comes to mind)
+
+We either **succeed** across **all** data stores or we **fail** and we have to be able to **recover** from that failure like if **nothing has happened**.
+
+
+The following diagram illustrates the high-level flow:
+
+```mermaid
+flowchart TD
+    NewBlock[A Block was just committed\nor we are at Height 0] -->CreateSavepoint(Create SavePoint)
+    CreateSavepoint --> UpdatingState(We are creating\na new block.\nUpdating the State)
+    UpdatingState --> ShouldCommit
+    ShouldCommit{Should we commit?} --> |Yes| ReceivedOKFromDatastores{Have all\nthe datastores\ncommitted\nsuccessfully?}
+    ShouldCommit --> |No| NewBlock
+    ReceivedOKFromDatastores -->|Yes| Committed[Committed successfully] --> |Repeat with\nthe new Block| NewBlock
+    ReceivedOKFromDatastores -->|No, we have at least one failure| Rollback[Rollback] --> |Repeat with\nthe rolled back Block | NewBlock
+
+```
+
+## Definitions
+
+### Savepoints
+
+A savepoint (also referred as checkpoints/snapshots depending on the implementation) is either the beginning of database transaction (or distributed transaction) or some sort of artifact that was created right after a successful commit happened that allows to recreate a perfect copy of the state at the time it was created.
+
+### Rollbacks
+
+A rollback is the process of cleanly reverting the state of the node to a previous state that we have saved in a savepoint.
+
+### Minimum Viable Product
+
+After having examined the `Persistence` and `Utility` modules, I have identified the following areas that we can consider as part of the MVP, these could be used as individual Github issues/deliverables potentially:
+
+- [**State changes invalidation and rollback triggering**] We need some sort of shared and thread-safe reference that is available for us across the whole call-stack that we can use in the event of a failure to flag that we need to abort whatever we are doing and rollback. This could be achieved via the use of the [context](https://pkg.go.dev/context) package.
```

Comment:
Fair point. I was initially thinking about embedding the go `Context` into the `PersistenceContext` and in my research I bumped into this: 

![image](https://user-images.githubusercontent.com/29378614/222203512-3654c22e-f25c-4270-8ebe-0c72f34c9d24.png)

source: https://go.dev/blog/context-and-structs

Reply:
This is interesting and makes a ton of sense.

I completely agree that having `Context` be in a struct that also acts as a receiver is THE WRONG approach.

The very first thing we did though had a struct called `PocketContext` that contained golang `Context` inside of it, which got passed around everywhere (i.e. starting with `node.go`) as a first param. 

Don't have a strong opinion if we should just use `Context` everywhere though. It seems to be what everyone else is doing.

Reply:
Devil's advocate: What if what everyone else is doing is wrong because they are all copying each other? (I should be spending more time looking at other projects, unfortunately I don't have good or bad examples, just assuming)

Or maybe sometimes a pattern makes sense but when applied it can get "lost in translation".

It makes me think about this:

![image](https://user-images.githubusercontent.com/29378614/223677621-bdeb987c-0e00-426f-96e1-3a069f6a3e2d.png)

Just kidding. 

The context is usually used to define the lifetime boundaries of what normally is an HTTP request.

In our case the "request" is effectively a "tentative world state transition application to the local state subject to byzantine consensus and unexpected errors".

Reply:
I had to tweet this: https://twitter.com/olshansky/status/1644090374242332672